server/src/main/java/org/elasticsearch/transport/OutboundHandler.java

```diff
@@ -116,7 +117,8 @@ private void sendMessage(TcpChannel channel, OutboundMessage networkMessage, Act
 
     private void internalSend(TcpChannel channel, SendContext sendContext) {
         channel.getChannelStats().markAccessed(threadPool.relativeTimeInMillis());
-        try {
+        // stash thread context so that channel event loop is not polluted by thread context
+        try (ThreadContext.StoredContext existing = threadPool.getThreadContext().stashContext()) {
```

Comment:
Do we even need to do this? Now that we fixed `NetworkMessage` I think there are only two possible scenarios here:

1. We are running this on the IO thread for the channel we send to. In that case, the context was polluted elsewhere already and this isn't really safe in the first place (because under load the actions in Netty's `flush` might run after this block executed and restored the unclean context for example)? 
2. We are running this off of the network thread of the channel, in which case switching the context here is just irrelevant?

Reply:
This is specifically for case 1, yes. We're handing control over to the `Netty` library in this case, which, as you noted, can execute under a non-default thread context (e.g. in case where a message was received on the same thread that is now executing under SAME). We just have to make sure that whenever give control back to the Netty code (which is free to write/flush at its own pace) that we clean up the context. If `Netty` decides running this later, that's fine, any time we hand control to Netty code, we're making sure to clean up the context.

Reply:
Makes sense thanks!